From fe5a6c2111288767e447262c512d23680c3e410e Mon Sep 17 00:00:00 2001
From: Don Zickus <dzickus@redhat.com>
Date: Fri, 12 Sep 2014 18:12:40 -0400
Subject: [powerpc] numa: Enable USE_PERCPU_NUMA_NODE_ID

Message-id: <1410545655-205645-532-git-send-email-dzickus@redhat.com>
Patchwork-id: 94426
O-Subject: [RHEL7 PATCH 531/626] powerpc/numa: Enable USE_PERCPU_NUMA_NODE_ID
Bugzilla: 1127366
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Steve Best <sbest@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>
RH-Acked-by: David Gibson <dgibson@redhat.com>

https://bugzilla.redhat.com/show_bug.cgi?id=1127366

commit 8c272261194dfda11cc046fbe808e052f6f284eb
Author: Nishanth Aravamudan <nacc@linux.vnet.ibm.com>
Date:   Mon May 19 11:14:23 2014 -0700

    powerpc/numa: Enable USE_PERCPU_NUMA_NODE_ID

    Based off 3bccd996 for ia64, convert powerpc to use the generic per-CPU
    topology tracking, specifically:

        initialize per cpu numa_node entry in start_secondary
        remove the powerpc cpu_to_node()
        define CONFIG_USE_PERCPU_NUMA_NODE_ID if NUMA

    Signed-off-by: Nishanth Aravamudan <nacc@linux.vnet.ibm.com>
    Signed-off-by: Benjamin Herrenschmidt <benh@kernel.crashing.org>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/arch/powerpc/Kconfig b/arch/powerpc/Kconfig
index ca7f90a..42511c2 100644
--- a/arch/powerpc/Kconfig
+++ b/arch/powerpc/Kconfig
@@ -455,6 +455,10 @@ config NODES_SHIFT
  default "4"
  depends on NEED_MULTIPLE_NODES
 
+config USE_PERCPU_NUMA_NODE_ID
+ def_bool y
+ depends on NUMA
+
 config ARCH_SELECT_MEMORY_MODEL
  def_bool y
  depends on PPC64
diff --git a/arch/powerpc/include/asm/topology.h b/arch/powerpc/include/asm/topology.h
index c778ca0..938ad4a 100644
--- a/arch/powerpc/include/asm/topology.h
+++ b/arch/powerpc/include/asm/topology.h
@@ -16,19 +16,6 @@ struct device_node;
 
 #include <asm/mmzone.h>
 
-static inline int cpu_to_node(int cpu)
-{
- int nid;
-
- nid = numa_cpu_lookup_table[cpu];
-
- /*
-  * During early boot, the numa-cpu lookup table might not have been
-  * setup for all CPUs yet. In such cases, default to node 0.
-  */
- return (nid < 0) ? 0 : nid;
-}
-
 #define parent_node(node) (node)
 
 #define cpumask_of_node(node) ((node) == -1 ?    \
diff --git a/arch/powerpc/kernel/smp.c b/arch/powerpc/kernel/smp.c
index e2a4232..d7252ad 100644
--- a/arch/powerpc/kernel/smp.c
+++ b/arch/powerpc/kernel/smp.c
@@ -390,6 +390,7 @@ void smp_prepare_boot_cpu(void)
 #ifdef CONFIG_PPC64
  paca[boot_cpuid].__current = current;
 #endif
+ set_numa_node(numa_cpu_lookup_table[boot_cpuid]);
  current_set[boot_cpuid] = task_thread_info(current);
 }
 
@@ -750,6 +751,11 @@ void start_secondary(void *unused)
  }
  traverse_core_siblings(cpu, true);
 
+ /*
+  * numa_node_id() works after this.
+  */
+ set_numa_node(numa_cpu_lookup_table[cpu]);
+
  smp_wmb();
  notify_cpu_starting(cpu);
  set_cpu_online(cpu, true);
-- 
1.7.1